### DISCUSSION TITLE
感觉 Caddy 的 SNI 分流比 Nginx 的 SNI 分流性能更好

### DISCUSSION THREAD

**Discussion by choicky** [lxhao61/integrated-examples#186] [Show and tell]
这两天做了一下简单的测试，SNI 分流场景有如下五种：
 1. [Xray(M+F+H+G+B+A)+Caddy(N)](https://github.com/lxhao61/integrated-examples/tree/main/Xray(M%2BF%2BH%2BG%2BB%2BA)%2BCaddy(N))（由 Caddy 兼顾 SNI 分流的综合应用。）
 2.  [tls-shunt-proxy](https://github.com/liberal-boy/tls-shunt-proxy) 前置分流
 3.  [Xray(E+F+H+G+B+A)+Caddy(N)+Nginx](https://github.com/lxhao61/integrated-examples/tree/main/Xray(E%2BF%2BH%2BG%2BB%2BA)%2BCaddy(N)%2BNginx)（以 XTLS Vision 为核心的综合应用。）
增加 `so_keepalive=on`参数
 4.  [Xray(M+F+H+G+B+A)+Caddy(N)+Nginx](https://github.com/lxhao61/integrated-examples/tree/main/Xray(M%2BF%2BH%2BG%2BB%2BA)%2BCaddy(N)%2BNginx)（以 REALITY Vision 为核心的综合应用。）
 5. [Xray(E+F+H+G+B+A)+Caddy(N)](https://github.com/lxhao61/integrated-examples/tree/main/Xray(E%2BF%2BH%2BG%2BB%2BA)%2BCaddy(N))（由 Caddy 兼顾 SNI 分流的综合应用。）

每个场景访问 https://fast.com 测速四轮，结果如下：
![屏幕截图 2023-10-02 123858](https://github.com/lxhao61/integrated-examples/assets/10532949/013c2597-0b0b-49fd-afbc-95552f4fd444)

虽然这种速度不是很科学，但某种程度上也具有一定的参考意义吧。根据这几轮测速，`caddy` l4分流性能最好；其次是 `tls-shunt-proxy`，而 `nginx`性能最差。坊间一直说 `nginx` 性能比 `caddy` 好，不知道是否还有优化的空间呢？

**Comment by lxhao61** [comment]


**Reply by choicky** [reply]
我的xray 服务配置了几个 inbounds，有 XTLS Vision，也有 REALITY Vision，都是 UDS 。

我喜欢用 xtls，所以测试时都是使用 xtls （五轮测试都是xtls，只是我可能原贴写错了配置文件的地址）。

1与5速度有差异、3与4速度有差异，有可能是网络波动或者vps厂商在调控？ 但我日常使用时，是有明显感受到 nginx分流速度不如caddy分流的。xray那边也有人讨论nginx分流会降速： https://github.com/XTLS/Xray-core/issues/685 这个帖子中有人说增加 `so_keepalive=on` 能改善网速，但我测试没发现有显著变化。

我手头没有比较稳定的服务器。你有条件时，可以测试一下caddy 与 nginx 的分流速度对比。

**Reply by lxhao61** [reply]
哦，若仅 XTLS Vision（或 REALITY Vision）且都是 UDS，那么仅有三种场景。
另外实际使用中我也发现 Caddy SNI 分流比 Nginx SNI 分流性能更好。

**Reply by choicky** [reply]
做第五轮测试，主要是排除vps厂商和宽带运营商（可能存在的）控速的影响。

第三轮与第四轮，主要是想看看网友说的参数  `so_keepalive=on` 是否有效。

**Reply by lxhao61** [reply]
哦，了解了。
Caddy 目前就是不支持 UDP 端口定向代理，否则就完美了。

**Reply by lxhao61** [reply]
Caddy（caddy-l4 最新版）已支持定向 UDP 转发了。

**Reply by choicky** [reply]
> Caddy（caddy-l4 最新版）已支持定向 UDP 转发了。

多谢通知。
参考你的配置 [integrated-examples](https://github.com/lxhao61/integrated-examples/tree/main)/[Xray(M+F+H+G+B+A)+Caddy(N)](https://github.com/lxhao61/integrated-examples/tree/main/Xray(M%2BF%2BH%2BG%2BB%2BA)%2BCaddy(N)) 用上了。暂时没发现问题。

**Reply by lxhao61** [reply]
你有空可以再测试如下性能（SNI 分流及 HTTP/3）：
1、由 Caddy 兼顾 SNI 分流实现 XTLS Vision 与 Trojan 回落为核心的综合应用
[Xray(E+F+H+G+B+A)+Caddy(N)](https://github.com/lxhao61/integrated-examples/tree/main/Xray(E%2BF%2BH%2BG%2BB%2BA)%2BCaddy(N))（由 Caddy 兼顾 SNI 分流的综合应用。）
2、由 Caddy 兼顾 SNI 分流实现 REALITY Vision 与 Trojan 回落为核心的综合应用
[Xray(M+F+H+G+B+A)+Caddy(N)](https://github.com/lxhao61/integrated-examples/tree/main/Xray(M%2BF%2BH%2BG%2BB%2BA)%2BCaddy(N))（由 Caddy 兼顾 SNI 分流的综合应用。）
3、由 Nginx 专职 SNI 分流实现兼顾各方优势的综合应用
[Xray(E+F+H+G+B+A)+Caddy(N)+Nginx](https://github.com/lxhao61/integrated-examples/tree/main/Xray(E%2BF%2BH%2BG%2BB%2BA)%2BCaddy(N)%2BNginx)（以 XTLS Vision 为核心的综合应用。）
[Xray(M+F+H+G+B+A)+Caddy(N)+Nginx](https://github.com/lxhao61/integrated-examples/tree/main/Xray(M%2BF%2BH%2BG%2BB%2BA)%2BCaddy(N)%2BNginx)（以 REALITY Vision 为核心的综合应用。）

**Reply by choicky** [reply]
趁凌晨网络比较好，同样用 https://fast.com 做了10组测试，如下图。

- caddy 的 sni 分流比nginx sni分流略快一点，但两者的差异远远小于我去年的测试。
-  caddy 与 nginx 的udp分流可能没有性能差异，或者是运营商丢我的包？
-  XTLS Vision 比 REALITY Vision 略快，但差异很小，在0%~8%左右；
-  REALITY Vision 比 ws 略快一点 （都是先 sni 分流的情况下）


<img width="478" alt="屏幕截图 2024-06-08 031107" src="https://github.com/lxhao61/integrated-examples/assets/10532949/399424b1-816e-4879-bdb1-006faca9208b">

**Reply by lxhao61** [reply]
感谢重新提供的测试分享。
就测试来看接近正常估计，但还是对 Nginx SNI 性能失望，毕竟 Caddy 也支持定向 UDP 转发后还是干不过，无一点优势了。
对了，你上边测试全部采用的配置 1 模式测试，还是采用的配置 2 模式测试？

**Reply by choicky** [reply]
今天中午发现网速比凌晨还好，所以我重新又做了一轮测试 （昨晚白熬夜了T_T）

#### 测试环境（服务器端）：
- Debian GNU/Linux trixie/sid (6.5.0-1-arm64)
- caddy v2.8.4
- nginx 1.22.1
- xray 1.8.13
- XTLS Vision 是域名x，caddy/nginx sni分流后转到UDS（模式2），xray监听UDS（模式2）
- REALITY Vision 是域名r，也是 caddy/nginx sni分流后转到UDS（模式2），xray监听UDS（模式2）
- vless tls ws 是域名w，caddy/nginx sni分流后丢给caddy 监听的 https UDS（模式2），然后 https UDS再丢给xray监听的UDS（模式2）
- hysteria 2 本身是监听端口，今天中午测试了直连、caddy/nginx UDP转发三种情形

今天中午的测试结果，有些数据支持之前的结论，有些不支持了（可能网络波动大？） 。

#### 今天中午的大概结论是：


-  XTLS Vision 模式下，caddy 的 sni 分流比nginx sni分流略快一点，但两者的差异远远小于我去年的测试；该结论与昨晚相同。
- REALITY Vision 模式下，caddy 的sni 分流与 nginx sni分流互有胜负，可认为打平手；该结论与昨晚有变化。
- vless tls ws 模式下，caddy 的sni 分流与 nginx sni分流也是互有胜负，感觉波动性比 REALITY Vision 模式大；该结论与昨晚有变化。
- caddy 与 nginx 的udp分流也是互有胜负，但是，网速明显比直连 `hysteria 2` 慢很多，说明UDP转发性能损耗很大；该结论与昨晚有变化。
-  XTLS Vision 比 REALITY Vision 略快，但差异很小，在0%~8%左右；该结论与昨晚相同。
-  caddy的REALITY Vision 比 ws 略快一点，但 nginx的REALITY Vision 比与ws是互有胜负；该结论与昨晚相同。

总体上来说，caddy sni 分流性能不输nginx或者略胜nginx；caddy 与 nginx 的UDP转发性能目前都不太理想。

![屏幕截图 2024-06-08 130948](https://github.com/lxhao61/integrated-examples/assets/10532949/239116f0-9afe-4f6c-bdea-6888633df791)

**Reply by lxhao61** [reply]
很详细的测试，谢谢！
1、Nginx 目前最新稳定版为 1.26.1，你的 Nginx 使用的系统自带安装，还是自己编译安装的？
2、定向 UDP 转发效率应该不太好，启用目的：主要解决 Caddy 的 NaiveProxy 位于后端无法共用（统一） TCP/UDP 443 端口问题，无必要转发给 Hysteria 2 来共用 TCP/UDP 443 端口。
3、关闭定向 UDP 443 转发，Hysteria 2 直接配置 HTTP/3 使用；或定向 UDP 443 转发不关，Hysteria 2 换一个端口配置混淆使用（类似Hysteria 1 应用）。（已提供 Hysteria 2 更新示例）

**Reply by choicky** [reply]
nginx 是Debian 12（当前的stable）系统默认的版本，通过apt的方式安装的。
最近看了一下，Debian Testing默认的nginx才是 `1.26.0` 。我周末在测试一下看看

**Comment by choicky** [comment]
另外，我折腾 `tls-shunt-proxy` 与 `nginx` 分流，本来是打算把 sni分流从 `caddy` 中剥离出来，让 `caddy` 能使用传统的 `Caddyfile` 的，因为我有还有几个网站，写 `json` 配置格式的文件很麻烦。——我有一个 `wordpress` ，一直没搞定 `caddy.json` 的配置

但如果 另外两者的sni分流性能比 caddy 差那么多的话，只能咬咬牙继续用 caddy.json 配置文件了。

**Reply by lxhao61** [reply]
给你一个笨办法：你先用 Caddyfile 配置无 SNI 部分，再使用命令：caddy adapt --config /path/to/Caddyfile --pretty 转换为 JSON 配置，再加上 SNI 部分配置即可。
